interfaces/repo.go

```diff
@@ -327,68 +327,182 @@ func (r *Repository) Connect(ref ConnRef) error {
+		conns, snaps, err = r.disconnectPlugFromSlot(plugSnapName, plugName, slotSnapName, slotName)
 	}
 
+	if err != nil {
+		return nil, nil, err
+	}
+	// Flatten map of snaps into a sorted list
+	names := make([]string, 0, len(snaps))
+	for snapName := range snaps {
+		names = append(names, snapName)
+	}
+	sort.Strings(names)
+	result := make([]*snap.Info, 0, len(snaps))
+	for _, snapName := range names {
+		result = append(result, snaps[snapName])
+	}
+	return conns, result, nil
 }
 
-// disconnectEverythingFromSnap finds a specific snap and disconnects all the plugs connected to all the slots therein.
-func (r *Repository) disconnectEverythingFromSnap(slotSnapName string) error {
-	if _, ok := r.slots[slotSnapName]; !ok {
-		return fmt.Errorf("cannot disconnect plug from snap %q, no such snap", slotSnapName)
+// ResolveDisconnectAll finds all of the connections that would be disconnected by DisconnectAll()
+func (r *Repository) ResolveDisconnectAll(snapName, plugOrSlotName string) ([]ConnRef, error) {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	return r.unlockedResolveDisconnectAll(snapName, plugOrSlotName)
+}
+
+func (r *Repository) unlockedResolveDisconnectAll(snapName, plugOrSlotName string) ([]ConnRef, error) {
+	if snapName == "" {
+		// Look up the core snap if no snap name was given
+		switch {
+		case r.slots["core"] != nil:
+			snapName = "core"
+		case r.slots["ubuntu-core"] != nil:
+			snapName = "ubuntu-core"
+		default:
+			return nil, fmt.Errorf("cannot resolve disconnect, snap name is empty")
+		}
 	}
-	for _, slot := range r.slots[slotSnapName] {
-		for plug := range r.slotPlugs[slot] {
+	var conns []ConnRef
+	if plugOrSlotName == "" {
+		// "wildcard" mode, affect all the plugs or slots in this snap
+		for _, plug := range r.plugs[snapName] {
+			for _, slotRef := range plug.Connections {
+				connRef := ConnRef{PlugRef: plug.Ref(), SlotRef: slotRef}
+				conns = append(conns, connRef)
+			}
+		}
+		for _, slot := range r.slots[snapName] {
+			for _, plugRef := range slot.Connections {
+				connRef := ConnRef{PlugRef: plugRef, SlotRef: slot.Ref()}
+				conns = append(conns, connRef)
+			}
+		}
+	} else {
+		// Precise mode, affect specific plug or slot in this snap
+		if plug, ok := r.plugs[snapName][plugOrSlotName]; ok {
+			for _, slotRef := range plug.Connections {
+				connRef := ConnRef{PlugRef: plug.Ref(), SlotRef: slotRef}
+				conns = append(conns, connRef)
+			}
+		}
+		if slot, ok := r.slots[snapName][plugOrSlotName]; ok {
+			for _, plugRef := range slot.Connections {
+				connRef := ConnRef{PlugRef: plugRef, SlotRef: slot.Ref()}
+				conns = append(conns, connRef)
+			}
+		}
+		// Check if plugOrSlotName actually maps to anything
+		if len(conns) == 0 {
+			return nil, fmt.Errorf("snap %q has no plug or slot named %q", snapName, plugOrSlotName)
+		}
+	}
+	return conns, nil
+}
+
+// DisconnectAll disconnects all of the given connections, returning the list of affected snap names.
```

Comment:
Why are we returning a list of snap names?  We asked the repository itself to resolve the list of affected connections for us, so we already have it at hand are actually handing it in. Feels strange to get this list out from this function.